### PR TITLE
replace js annotation typeclass method

### PR DESCRIPTION
**Pull request by chikeabuah**
still doesn't seem to solve my particular issue, but i could have missed something.

### CODE REVIEW

File: hs/src/Reach/JSUtil.hs [hunk 1]
```diff
@@ -267,19 +320,63 @@ instance HasJSAnnot JSExpression where
...
     JSYieldFromExpression a _ _ -> a
+  rjsa a' = \case
+    JSIdentifier _ x -> JSIdentifier a' x
+    JSDecimal _ x -> JSDecimal a' x
+    JSLiteral _ x -> JSLiteral a' x
+    JSHexInteger _ x -> JSHexInteger a' x
+    JSOctal _ x -> JSOctal a' x
+    JSStringLiteral _ x -> JSStringLiteral a' x
+    JSRegEx _ x -> JSRegEx a' x
+    JSArrayLiteral _ x _ -> JSArrayLiteral a' x a'
```
Comment (jeapostrophe) [inline hs/src/Reach/JSUtil.hs:331]: `x` has annotations inside it

File: hs/src/Reach/JSUtil.hs [hunk 2]
```diff
@@ -267,19 +320,63 @@ instance HasJSAnnot JSExpression where
...
+    JSAssignExpression a x y -> JSAssignExpression (rjsa a' a) (rjsa a' x) (rjsa a' y)
+    JSAwaitExpression _ x -> JSAwaitExpression a' (rjsa a' x)
+    JSCallExpression a _ y _ -> JSCallExpression (rjsa a' a) a' (rjsaJSL a' y) a'
+    JSCallExpressionDot a _ y -> JSCallExpressionDot (rjsa a' a) a' (rjsa a' y)
+    JSCallExpressionSquare a _ y _ -> JSCallExpressionSquare (rjsa a' a) a' (rjsa a' y) a'
+    JSClassExpression _ x y z r _ -> JSClassExpression a' x y z r a'
```
Comment (jeapostrophe) [inline hs/src/Reach/JSUtil.hs:337]: If we ever support classes, we're going to forget about this

File: hs/src/Reach/JSUtil.hs [hunk 3]
```diff
@@ -267,19 +320,63 @@ instance HasJSAnnot JSExpression where
...
+    JSCommaExpression a _ y -> JSCommaExpression (rjsa a' a) a' (rjsa a' y)
+    JSExpressionBinary a x y -> JSExpressionBinary (rjsa a' a) (rjsa a' x) (rjsa a' y)
+    JSExpressionParen _ x _ -> JSExpressionParen a' (rjsa a' x) a'
+    JSExpressionPostfix a x -> JSExpressionPostfix (rjsa a' a) (rjsa a' x)
+    JSExpressionTernary a _ y _ r -> JSExpressionTernary (rjsa a' a) a' (rjsa a' y) a' (rjsa a' r)
+    JSArrowExpression x _ y -> JSArrowExpression x a' y
```
Comment (jeapostrophe) [inline hs/src/Reach/JSUtil.hs:343]: x and y both have stuff in them and almost certainly are being used in your case

File: hs/src/Reach/JSUtil.hs [hunk 4]
```diff
@@ -267,19 +320,63 @@ instance HasJSAnnot JSExpression where
...
+    JSFunctionExpression _ x _ z _ s -> JSFunctionExpression a' x a' (rjsaJSL a' z) a' s
```
Comment (jeapostrophe) [inline hs/src/Reach/JSUtil.hs:344]: You should look into `s`

File: hs/src/Reach/JSUtil.hs [hunk 5]
```diff
@@ -267,19 +320,63 @@ instance HasJSAnnot JSExpression where
...
+    JSGeneratorExpression _ _ y _ r _ t -> JSGeneratorExpression a' a' y a' (rjsaJSL a' r) a' t
```
Comment (jeapostrophe) [inline hs/src/Reach/JSUtil.hs:345]: And `t`

File: hs/src/Reach/JSUtil.hs [hunk 6]
```diff
@@ -267,19 +320,63 @@ instance HasJSAnnot JSExpression where
...
+    JSMemberDot a _ y -> JSMemberDot (rjsa a' a) a' (rjsa a' y)
+    JSMemberExpression a _ y _ -> JSMemberExpression (rjsa a' a) a' (rjsaJSL a' y) a'
+    JSMemberNew _ x _ z _ -> JSMemberNew a' (rjsa a' x) a' (rjsaJSL a' z) a'
+    JSMemberSquare a _ y _ -> JSMemberSquare (rjsa a' a) a'  (rjsa a' y) a'
+    JSNewExpression _ x -> JSNewExpression a' (rjsa a' x)
+    JSObjectLiteral _ x _ -> JSObjectLiteral a' (mkCommaTrailingList (map (rjsa a') $ jso_flatten x)) a'
```
Comment (jeapostrophe) [inline hs/src/Reach/JSUtil.hs:351]: You should just implement this typeclass on this structure

File: hs/src/Reach/JSUtil.hs [hunk 7]
```diff
@@ -267,19 +320,63 @@ instance HasJSAnnot JSExpression where
...
+    JSSpreadExpression _ x -> JSSpreadExpression a' (rjsa a' x)
+    JSTemplateLiteral x _ y z -> JSTemplateLiteral (fmap (rjsa a') x) a' y z
```
Comment (jeapostrophe) [inline hs/src/Reach/JSUtil.hs:353]: Ditto and z has stuff in it

File: hs/src/Reach/JSUtil.hs [hunk 8]
```diff
@@ -267,19 +320,63 @@ instance HasJSAnnot JSExpression where
...
+    JSUnaryExpression a x -> JSUnaryExpression (rjsa a' a) (rjsa a' x)
+    JSVarInitExpression a x -> JSVarInitExpression (rjsa a' a) x
```
Comment (jeapostrophe) [inline hs/src/Reach/JSUtil.hs:355]: x is important to look at

File: hs/src/Reach/JSUtil.hs [hunk 9]
```diff
@@ -267,19 +320,63 @@ instance HasJSAnnot JSExpression where
...
+    JSObjectMethod jmd -> JSObjectMethod $ rjsa a' jmd
+    JSObjectSpread _ x -> JSObjectSpread a' (rjsa a' x)
 
 instance HasJSAnnot JSMethodDefinition where
   jsa = \case
     JSMethodDefinition _ ja _ _ _ -> ja
     JSGeneratorMethodDefinition ja _ _ _ _ _ -> ja
     JSPropertyAccessor _ _ ja _ _ _ -> ja
+  rjsa a' = \case
+    JSMethodDefinition x _ y _ r -> JSMethodDefinition x a' (rjsaJSL a' y) a' r
```
Comment (jeapostrophe) [inline hs/src/Reach/JSUtil.hs:377]: r

File: hs/src/Reach/JSUtil.hs [hunk 10]
```diff
@@ -267,19 +320,63 @@ instance HasJSAnnot JSExpression where
...
+    JSGeneratorMethodDefinition _ x _ z _ s -> JSGeneratorMethodDefinition a' x a' (rjsaJSL a' z) a' s
```
Comment (jeapostrophe) [inline hs/src/Reach/JSUtil.hs:378]: s

File: hs/src/Reach/JSUtil.hs [hunk 11]
```diff
@@ -267,19 +320,63 @@ instance HasJSAnnot JSExpression where
...
+    JSPropertyAccessor x y _ z _ s -> JSPropertyAccessor x y a' (rjsaJSL a' z) a' s
```
Comment (jeapostrophe) [inline hs/src/Reach/JSUtil.hs:379]: s

File: hs/src/Reach/JSUtil.hs [hunk 12]
```diff
@@ -319,6 +416,43 @@ instance HasJSAnnot JSStatement where
     JSVariable ja _ _ -> ja
     JSWhile ja _ _ _ _ -> ja
     JSWith ja _ _ _ _ _ -> ja
+  rjsa a' = \case
+    JSStatementBlock _ x _ z -> JSStatementBlock a' (fmap (rjsa a') x) a' z
```
Comment (jeapostrophe) [inline hs/src/Reach/JSUtil.hs:420]: You'll need to go inside the semis too (z)

File: hs/src/Reach/JSUtil.hs (hunk 8 above)
Comment (chikeabuah) [reply hs/src/Reach/JSUtil.hs:355]: why isn't x currently a part of this typeclass then?

Comment (jeapostrophe) [reply hs/src/Reach/JSUtil.hs:355]: Cause I'm lazy :)

**Pull request closed without merging**